File: server/routes/slides.js
const Slide = require('../models').Slide;

module.exports = function(app) {

    app.get('/slides', (req, res) => {
        Slide.findAll().then(slides => res.json(slides));
    });

    app.get('/slides/:id', (req, res) => {
        Slide.findById(req.params.id).then(slide => {
            if (!slide) {
                res.status(404).send();
                return;
            }

            res.json(slide.toJSON());
        });
    });

    app.post('/slides', (req, res) => {
        Slide.create({
            title: req.body.title,
            body: req.body.body,
            visible: req.body.visible,
            type: req.body.type,
            index: req.body.index,
            name: req.body.name
        }).then((instance) => {
            if (!instance) {
                res.status(500).send();
                return;
            }
            res.json(instance.toJSON());
        }).catch(() => {
            res.status(500).send();
        });
    });

    app.put('/slides/:id', (req, res) => {
        Slide.findById(req.params.id).then(slide => {
            if (!slide) {
                res.status(404).send();
                return;
            }

            slide.update({
                title: req.body.title,
                body: req.body.body,
                visible: req.body.visible,
                index: req.body.index,
                name: req.body.name
            }).then(newSlide => {
                res.json(newSlide.toJSON());
            });
        }).catch(() => {
            res.status(500).send();
        });
    });

    app.delete('/slides/:id', (req, res) => {
        Slide.findById(req.params.id).then(slide => {
            if (!slide) {
                res.status(404).send();
                return;
            }
            slide.destroy().then(() => {
                res.status(200).send();
            }).catch(() => {
                res.status(500).send();
            });
        }).catch(() => {
            res.status(500).send();
        });
    });
};

File: java-backend/src/main/java/no/javazone/switcharoo/api/Slides.java
package no.javazone.switcharoo.api;

import com.google.gson.Gson;
import no.javazone.switcharoo.Authentication;
import no.javazone.switcharoo.api.mapper.SlideMapper;
import no.javazone.switcharoo.api.model.Slide;
import no.javazone.switcharoo.dao.SlidesDao;
import no.javazone.switcharoo.exception.BadRequestException;
import no.javazone.switcharoo.exception.NotFoundException;

import static no.javazone.switcharoo.Util.*;
import static no.javazone.switcharoo.api.verifier.SlideVerifier.verify;
import static spark.Spark.*;

public class Slides implements HttpService {

    private final SlidesDao slides;
    private final Authentication auth;

    public Slides(SlidesDao slides, Authentication auth) {
        this.slides = slides;
        this.auth = auth;
    }

    @Override
    public void register(Gson gson) {
        path("/", () -> {
            get("/slides",
                (req, res) -> slides.list().map(SlideMapper::fromDb).toJavaList(), gson::toJson);

            get("/slides/:id",
                (req, res) -> parseLong(req.params(":id"))
                    .flatMap(id -> slides.get(id))
                    .map(SlideMapper::fromDb)
                    .getOrElseThrow(NotFoundException::new),
                gson::toJson
            );

            post("/slides",
                (req, res) -> verify(gson.fromJson(req.body(), Slide.class))
                    .map(SlideMapper::toDb)
                    .flatMap(slides::create)
                    .map(SlideMapper::fromDb)
                    .getOrElseThrow(BadRequestException::new),
                gson::toJson
            );

            put("/slides/:id",
                (req, res) -> verify(gson.fromJson(req.body(), Slide.class))
                    .map(SlideMapper::toDb)
                    .flatMap(s -> parseLong(req.params(":id")).map(id -> s.withId(id)))
                    .flatMap(slides::update)
                    .map(SlideMapper::fromDb)
                    .getOrElseThrow(BadRequestException::new),
                gson::toJson
            );

            delete("/slides/:id",
                (req, res) -> parseInt(req.params(":id"))
                    .flatMap(slides::delete)
                    .map(deleted -> {
                        if (deleted) {
                            res.status(200);
                        } else {
                            res.status(404);
                        }
                        return "";
                    })
                    .getOrElseThrow(BadRequestException::new),
                gson::toJson
            );

            before("/slides", (req, res) -> { if (!auth.verify(req)) halt(401);});
            before("/slides/*", (req, res) -> { if (!auth.verify(req)) halt(401);});
            after("/slides", (req, res) -> res.type("application/json"));
            after("/slides/*", (req, res) -> res.type("application/json"));
        });

    }
}
